File: python_port/rag/impl/simple_reranker.py
from typing import List
from ..interfaces import Reranker, Hit

class SimpleReranker(Reranker):
    def rerank(self, query_terms: List[str], hits: List[Hit]) -> List[Hit]:
        for hit in hits:
            proximity_bonus = self.calculate_proximity_bonus(hit.text, query_terms)
            title_boost = self.calculate_title_boost(hit.docId, query_terms)
            
            hit.score = (hit.score * 10) + proximity_bonus + title_boost
        
        # Sort by score DESC
        hits.sort(key=lambda h: h.score, reverse=True)
        return hits

    def calculate_proximity_bonus(self, text: str, terms: List[str]) -> float:
        # Simplified implementation as per Java code
        return 0.0

    def calculate_title_boost(self, title: str, terms: List[str]) -> float:
        lower_title = title.lower()
        for term in terms:
            if term in lower_title:
                return 3.0
        return 0.0

File: Iteration1CSE3063Proje/src/main/java/com/cse3063/rag/impl/SimpleReranker.java
package com.cse3063.rag.impl;

import com.cse3063.rag.interfaces.Reranker;
import com.cse3063.rag.interfaces.Retriever.Hit;
import java.util.List;
import java.util.stream.Collectors;

public class SimpleReranker implements Reranker {

    @Override
    public List<Hit> rerank(List<String> queryTerms, List<Hit> hits) {
        // Apply bonus scores
        for (Hit hit : hits) {
            double proximityBonus = calculateProximityBonus(hit.text, queryTerms);
            double titleBoost = calculateTitleBoost(hit.docId, queryTerms); // Simplified: check docId as title

            hit.score = (hit.score * 10) + proximityBonus + titleBoost;
        }

        // Sort by score DESC
        hits.sort((h1, h2) -> Double.compare(h2.score, h1.score));
        return hits;
    }

    private double calculateProximityBonus(String text, List<String> terms) {
        // +5 if any two query terms appear within 15 characters
        // Simplified implementation
        return 0.0;
    }

    private double calculateTitleBoost(String title, List<String> terms) {
        // +3 if doc title contains any query term
        for (String term : terms) {
            if (title.toLowerCase().contains(term)) {
                return 3.0;
            }
        }
        return 0.0;
    }
}
